### PR TITLE
Allow ExternalGraphic image to be used for GetCapabilities and GetLegendGraphics

### PR DESCRIPTION
**Pull request by jodygarnett**
This is a replacement for #400 adding LegendInfo to StyleInfo for use in WMS GetCapabilities generation. The data structure (if available) is also used in GetLegendGraphics.

* http://jira.codehaus.org/browse/GEOS-6425
* http://geoserver.org/display/GEOS/GSIP+111+GetLegendGraphicsRequest+Cleanup

The scope of this feature has been refined to restrict these external graphic references to images that are being served by GeoServer (for example from the styles directory). This allows GeoServer to still be responsible for all the information it is publishing, and ensure that the images are available for the GetLegendGraphics operation.

Update: Due to feedback from @aaime the above restriction is lifted, a warning will be provided in the user interface if you configure GeoServer with an absolute URL pointing to an external server. An absolute URL that starts with the base url is fine. During GetLegendGraphics any online resource that cannot be resolved will result in a warning  in the logs, and the code will proceed with dynamic generation.

To do:

* [x] Update ExternalGraphicPanel to restrict graphics to images in the styles directory
* [x] Submit proposal to cleanup GetLegendGraphicsRequest (GSIP 111)
* [x]  Recording images in GetLegendGraphicRequest using resource name, change to use of style name to allow for the same data with two different styles
* [x] Resolve images in GetLegendGraphicRequest as local file references (KVP parser has access to data directory) to allow local test cases to pass
* [x] Test GetCapabilities
* [x] Test GetLegendGraphic with LAYER and STYLE
* [x] Test GetLegendGraphic with LAYER and STYLE (rescale)
* [x] Test GetLegendGraphic with with no style (i.e. ensure default style picked up)
* [x] Test GetLegendGraphic with with no layer (i.e. used as a style preview)
* [x] Test GetLegendGraphic with group layer
* [x] Update user docs
* [x] Issue: Prepending of styles directory (see   http://localhost:8080/geoserver/styles/http://visual.merriam-webster.com/images/earth/meteorology/station-model.jpg)
* [x] Issue: Warnings not cleared in wicket form
* [x] Issue: Validation not run on form submit (only when "update" link is clicked).
* [x] Issue: Confirm assignment agreement for #400

### CODE REVIEW

**Comment by aaime** [comment]
So... I guess this one will be merged for 2.8.x?

**Comment by jodygarnett** [comment]
Unless I get to it tomorrow :) Need to rebase and see if it is any good...

**Comment by aaime** [comment]
2.8.x ready to receive this (just sayin... :-p )

**Comment by aaime** [comment]
Another couple of months passed since last comment. Still in the pipeline for being worked on?

**Comment by jodygarnett** [comment]
Yeah I asked Travis to take it to a branch so we could both look at it:
https://github.com/boundlessgeo/geoserver/tree/legend-capabilities

Near as I can tell he got stuck in wicket code.

--
Jody Garnett

**Comment by aaime** [comment]
Any relationship between this pull request and this ticket?
 https://osgeo-org.atlassian.net/browse/GEOS-7160

Maybe they are similar but not same, cannot decide :)

**Comment by jodygarnett** [comment]
I think this feature (ability to parse out a LegendGraphic from SLD) may in
fact be covered.

--
Jody Garnett

**Comment by jodygarnett** [comment]
@tbarsballe thanks for the help on this one, can you review the checklist at the top of this pull request and cross off them items that were fixed?

I think we may be done?!

**Comment by tbarsballe** [comment]
Of the remaining unchecked issues, I did not explicitly fix (or test) any of them, although from general testing I think both "prepending of styles directory" and "validation on submit" are fixed now.

Other fixes:
* Wicket tests for Style Legend Graphic fixed - All the unit tests pass now
* Discard Legend button now works as expected
* Use buttons instead of links to show/hide the legend entry (for better consistency with the rest of the GeoServer UI.

**Comment by tbarsballe** [comment]
Looks like @jodygarnett added test cases for the remaining issues last week. Now that all the test cases pass, we can consider these issues fixed.

**Comment by tbarsballe** [comment]
https://github.com/geoserver/geoserver/pull/1237 broke the mergeability of this PR.

**Comment by tbarsballe** [comment]
Manualy squashed and merged to master: 83bb4d6ca9e1e20a196e81a419d82fafcc8e84d1

File: src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetLegendGraphicTest.java [hunk 1]
```diff
@@ -78,6 +95,42 @@ public void testEnv() throws Exception {
...
+        // test rescale
+        base = "wms?service=WMS&version=1.1.1&request=GetLegendGraphic" +
+                "&layer=sf:states&style=custom" +
+                "&format=image/png&width=16&height=16";
+
+        image = getAsImage(base, "image/png");        
+        
+        String user_home = System.getProperty("user.home");
+        File home = new File(user_home);
+        ImageIO.write(expected, "png",  new File(home,"expected.png"));
```
Comment (bencaradocdavies) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetLegendGraphicTest.java]: @jodygarnett @tbarsballe please do not write test images to user home directory. Fixed in 438097734641e2d99a2346d186179b0bd0338e08 on master. These images were not used in the test and I assume that they were for test development and inadvertently included in this PR. At the least, ensure that test images are confined to module target so they can be cleaned. See gt-render for examples of how to write test fixtures that enable interactive development of tests and by default write no files.

Comment (jodygarnett) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetLegendGraphicTest.java]: Thanks Ben, sorry we missed that in such a long lived patch.
On Wed, Nov 4, 2015 at 4:16 AM Ben Caradoc-Davies <notifications@github.com>
wrote:

> In
> src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetLegendGraphicTest.java
> <https://github.com/geoserver/geoserver/pull/551#discussion_r43839044>:
>
> > +        BufferedImage image = getAsImage(base, "image/png");
> > +        Resource resource = getResourceLoader().get("styles/legend.png");
> > +        BufferedImage expected = ImageIO.read( resource.file() );
> > +
> > +        assertEquals( getPixelColor(expected,10,2).getRGB(), getPixelColor(image,10,2).getRGB() );
> > +
> > +        // test rescale
> > +        base = "wms?service=WMS&version=1.1.1&request=GetLegendGraphic" +
> > +                "&layer=sf:states&style=custom" +
> > +                "&format=image/png&width=16&height=16";
> > +
> > +        image = getAsImage(base, "image/png");
> > +
> > +        String user_home = System.getProperty("user.home");
> > +        File home = new File(user_home);
> > +        ImageIO.write(expected, "png",  new File(home,"expected.png"));
>
> @jodygarnett <https://github.com/jodygarnett> @tbarsballe
> <https://github.com/tbarsballe> please do not write test images to user
> home directory. Fixed in 4380977
> <https://github.com/geoserver/geoserver/commit/438097734641e2d99a2346d186179b0bd0338e08>
> on master. These images were not used in the test and I assume that they
> were for test development and inadvertently included in this PR. At the
> least, ensure that test images are confined to module target so they can be
> cleaned. See gt-render for examples of how to write test fixtures that
> enable interactive development of tests and by default write no files.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoserver/geoserver/pull/551/files#r43839044>.
>
-- 
--
Jody Garnett

Comment (bencaradocdavies) [inline src/wms/src/test/java/org/geoserver/wms/wms_1_1_1/GetLegendGraphicTest.java]: No worries, Jody.

On 04/11/15 17:11, Jody Garnett wrote:
> Thanks Ben, sorry we missed that in such a long lived patch.

-- 
Ben Caradoc-Davies <ben@transient.nz>
Director
Transient Software Limited <http://transient.nz/>
New Zealand

**Pull request closed without merging**